File: c/Examples/binary_search.c
#include <stdio.h>
int main()
{
    int arr[] = {23, 435, 65, 67, 71, 799, 787, 8, 7, 65, 60, 646, 4, 64, 6, 53, 2, 452, 45, 45, 642, 34, 14, 134, 14};
    int i;
    int j;
    int temp;
    int size;
    int low, mid, high;
    size = sizeof(arr) / sizeof(int);
    for (i = 0; i < size - 1; i++)
    {
        for (j = i + 1; j < size; j++)
        {
            if (arr[i] > arr[j])
            {
                temp = arr[i];
                arr[i] = arr[j];
                arr[j] = temp;
            }
        }
    }
    printf("%d\n", size);
    temp = 642;
    low = 0;
    high = size - 1;
    mid = (high + low) / 2;
    while (1)
    {
        mid = low + (high - low) / 2;
        if (arr[mid] == temp)
        {
            printf("%d --> %d", temp, mid);
            break;
        }
        else if (arr[mid] > temp)
        {
            high = mid - 1;
        }
        else if (arr[mid] < temp)
        {
            low = mid + 1;
        }
    }

    return 0;
}

File: Java/Examples/src/search/BinarySearch.java
package search;

public class BinarySearch {

    private static int binarySearch(int[] arr, int x, int low, int high) {
        if (low > high) {
            return -1;
        }
        int mid = low + (high - low) / 2;
        if (arr[mid] == x) {
            return mid;
        } else if (arr[mid] > x) {
            return binarySearch(arr, x, low, mid - 1);
        } else {
            return binarySearch(arr, x, mid + 1, high);
        }
    }

    private static int binarySearch(int[] arr, int x) {
        int low = 0;
        int high = arr.length - 1;
        while (low <= high) {
            int mid = low + (high - low) / 2;
            if (arr[mid] == x) {
                return mid;
            } else if (arr[mid] > x) {
                high = mid - 1;
            } else {
                low = mid + 1;
            }
        }
        return -1;
    }

    public static void main(String[] args) {
        int[] arr = { -999, -12, -3, 1, 1, 2, 3, 4, 6, 6, 6, 8, 9, 10, 23, 28, 41, 999 };
        System.out.println(binarySearch(arr, 999));
        System.out.println(binarySearch(arr, -12, 0, arr.length));
    }
}
